File: vol1/java-examples/src/main/java/com/heatonresearch/aifh/examples/learning/LearnXOR.java
/*
 * Artificial Intelligence for Humans
 * Volume 1: Fundamental Algorithms
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh

 * Copyright 2013 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */

package com.heatonresearch.aifh.examples.learning;

import com.heatonresearch.aifh.general.data.BasicData;
import com.heatonresearch.aifh.learning.RBFNetwork;
import com.heatonresearch.aifh.learning.TrainGreedyRandom;
import com.heatonresearch.aifh.learning.score.ScoreFunction;
import com.heatonresearch.aifh.learning.score.ScoreRegressionData;

import java.util.List;

/**
 * Learn the XOR function with a RBF Network trained by Greedy Random.
 */
public class LearnXOR extends SimpleLearn {

    /**
     * The input necessary for XOR.
     */
    public static final double[][] XOR_INPUT = {{0.0, 0.0}, {1.0, 0.0},
            {0.0, 1.0}, {1.0, 1.0}};

    /**
     * The ideal data necessary for XOR.
     */
    public static final double[][] XOR_IDEAL = {{0.0}, {1.0}, {1.0}, {0.0}};

    /**
     * Perform the example.
     */
    public void process() {
        final List<BasicData> trainingData = BasicData.convertArrays(XOR_INPUT, XOR_IDEAL);
        final RBFNetwork network = new RBFNetwork(2, 5, 1);
        final ScoreFunction score = new ScoreRegressionData(trainingData);
        final TrainGreedyRandom train = new TrainGreedyRandom(true, network, score);
        performIterations(train, 1000000, 0.01, true);
        query(network, trainingData);
    }

    /**
     * The main method.
     *
     * @param args Not used.
     */
    public static void main(final String[] args) {
        final LearnXOR prg = new LearnXOR();
        prg.process();
    }
}

File: vol1/csharp-examples/AIFH-Vol1/Examples/Learning/LearnXOR.cs
using AIFH_Vol1.Core.General.Data;
using AIFH_Vol1.Core.Learning;
using AIFH_Vol1.Core.Learning.Score;

namespace AIFH_Vol1.Examples.Learning
{
    /// <summary>
    /// Learn the XOR function using RBF network & Greedy Random algorithm
    /// </summary>
    public class LearnXOR : SimpleLearn
    {
        /// <summary>
        /// The name of this example.
        /// </summary>
        public static string ExampleName = "Learn the XOR function with a RBF Network trained by Greedy Random";

        /// <summary>
        /// The chapter this example is from.
        /// </summary>
        public static int ExampleChapter = 7;

        /// <summary>
        /// The input necessary for XOR.
        /// </summary>
        public static double[][] XorInput = {
            new [] {0.0, 0.0}, 
            new [] {1.0, 0.0},
            new [] {0.0, 1.0}, 
            new [] {1.0, 1.0}};

        /// <summary>
        /// The ideal data necessary for XOR.
        /// </summary>
        public static double[][] XorIdeal =
        {
            new [] {0.0}, 
            new [] {1.0}, 
            new [] {1.0}, 
            new [] {0.0}
        };

        /// <summary>
        /// Perform the example.
        /// </summary>
        public void Process()
        {
            var trainingData = BasicData.ConvertArrays(XorInput, XorIdeal);
            var network = new RBFNetwork(2, 5, 1);
            var score = new ScoreRegressionData(trainingData);
            var train = new TrainGreedyRandom(true, network, score);
            PerformIterations(train, 1000000, 0.01, true);
            Query(network, trainingData);
        }

        /// <summary>
        ///     The entry point for this example.  If you would like to make this example
        ///     stand alone, then add to its own project and rename to Main.
        /// </summary>
        /// <param name="args">Not used.</param>
        public static void ExampleMain(string[] args)
        {
            var prg = new LearnXOR();
            prg.Process();
        }
    }
}
